channel/player.go

```diff
@@ -1398,36 +1419,313 @@ func (d *player) loadAndApplyBuffSnapshot() {
 	defer rows.Close()
 
 	snaps := make([]BuffSnapshot, 0, 8)
+	toDelete := make([]int32, 0, 8)
+
 	now := time.Now().UnixMilli()
 	for rows.Next() {
 		var s BuffSnapshot
 		if err := rows.Scan(&s.SourceID, &s.Level, &s.ExpiresAtMs); err != nil {
 			log.Println("loadBuffSnapshot scan:", err)
 			return
 		}
-		if s.ExpiresAtMs > 0 && s.ExpiresAtMs <= now {
-			continue // skip already-expired
+
+		if s.ExpiresAtMs == 0 {
+			toDelete = append(toDelete, s.SourceID)
+			continue
 		}
+
+		normalized := s.ExpiresAtMs
+		if normalized > 0 && normalized < 1000000000000 {
+			normalized *= 1000
+		}
+
+		if normalized <= 0 || normalized <= now {
+			toDelete = append(toDelete, s.SourceID)
+			continue
+		}
+
+		s.ExpiresAtMs = normalized
 		snaps = append(snaps, s)
 	}
 	if err := rows.Err(); err != nil {
 		log.Println("loadBuffSnapshot rows:", err)
 		return
 	}
 
+	if len(toDelete) > 0 {
+		var b strings.Builder
+		b.WriteString("DELETE FROM character_buffs WHERE characterID=? AND sourceID IN (")
+		for i := range toDelete {
+			if i > 0 {
+				b.WriteByte(',')
```

Comment:
Would prefer hex codes rather than ascii unless the ascii representation contains meaning?